src/shell.cpp

```diff
@@ -0,0 +1,138 @@
+#include "shell.h"
+#include <iostream>
+#include <sstream>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+
+
+/**
+ * Runs the shell program, allowing the user to input commands and execute them.
+ * The shell supports running commands both in the foreground and background.
+ */
+void Shell::run() {
+    std::string input;
+
+    while (true) {
+        removeFinishedProcesses();
+
+        std::cout << "Shell> ";
+        std::getline(std::cin, input);
+
+        if (input == "exit") {
+            break;
+        }
+
+        if (input == "myjobs") {
+            showBackgroundProcesses();
+            continue;
+        }
+
+        // Check if the command should be run in the background
+        bool runInBackground = false;
+        if (!input.empty() && input.back() == '&') {
+            runInBackground = true;
+            input.pop_back();
+        }
+
+        executeCommand(input, runInBackground);
+    }
+}
+
+/**
+ * Executes the given command in a child process.
+ * If the command should be run in the background, it adds the background process
+ * to the list of background processes and prints a message.
+ * If the command should be run in the foreground, it waits for the child process to finish.
+ * @param command The command to execute
+ * @param runInBackground Flag indicating whether the command should be run in the background
+ */
+void Shell::executeCommand(const std::string& command, bool runInBackground) {
+    pid_t pid = fork();
+    if (pid == -1) {
+        std::cerr << "Failed to create child process" << std::endl;
+        return;
+    } 
+    
+    else if (pid == 0) {
+        // Child process
+        std::vector<std::string> args = splitString(command, ' ');
+
+        // Convert vector of strings to char* array
+        std::vector<char*> cArgs;
+        for (const auto& arg : args) {
+            cArgs.push_back(const_cast<char*>(arg.c_str()));
+        }
+        cArgs.push_back(nullptr); // Add a null terminator at the end
+
+        execvp(cArgs[0], cArgs.data());
```

Comment:
execvp finds the executable file for you, but you were asked to implement that by yourself